src/EditorFeatures/Core/Implementation/Formatting/FormattingInteractionServiceProxy.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    /// <summary>
+    /// Acts as an internal proxy for getting an <see cref="IFormattingInteractionService"/> or, if unavailable, an <see cref="IEditorFormattingService"/>.
+    /// This should be removed when <see cref="IEditorFormattingService"/> is fully removed.
+    /// </summary>
+    internal struct FormattingInteractionServiceProxy : ILanguageService
+    {
+        public static FormattingInteractionServiceProxy? GetService(Document document)
+        {
+            var formattingService = document.GetLanguageService<IFormattingInteractionService>();
+            IEditorFormattingService? editorService = null;
+
+            if (formattingService == null && (editorService = document.GetLanguageService<IEditorFormattingService>()) == null)
```

Comment:
oh god, my eyes.

Reply:
Just you wait until I have to add `#pragma warning suppress CS0618` and `#pragma warning restore CS0168` around it... :). Delete fast.

Reply:
➡️ Proxy is now removed.